File: wasm/src/providers/anthropic.ts
import type { RenderedContext, ToolSchema, StreamEvent, TextDelta, ThinkingDelta, ToolCallEvent, UsageEvent, LLMProvider, Message, ProviderDescriptor, ProviderReplay } from "../types.js"
import { assistantReplayKey, collectStreamMessage, toAnthropicMessages } from "./base.js"

/** Anthropic accepts at most this many cache_control breakpoints per request. */
const MAX_CACHE_BREAKPOINTS = 4
/** Rolling cache breakpoints reserved for the message history (system uses ≤2). */
const MESSAGE_CACHE_BREAKPOINTS = 2

function buildAnthropicTools(tools: ToolSchema[], anchorCache: boolean) {
  return tools.map((t, i) => ({
    name: t.name,
    description: t.description,
    input_schema: JSON.parse(t.parameters),
    // Anchor a tool breakpoint only when the system blocks won't carry one;
    // otherwise systemStable already caches the tools prefix (tools render
    // first), and a redundant tool breakpoint would burn a slot the message
    // history needs to stay within the 4-breakpoint budget.
    ...(anchorCache && i === tools.length - 1 ? { cache_control: { type: "ephemeral" as const } } : {}),
  }))
}

/**
 * Roll cache breakpoints across the conversation tail so the message-history
 * prefix is written once and re-read on later turns (without this the cached
 * prefix stops at the end of `system` and the whole tool-result history is
 * re-billed at full input price every turn).
 *
 * When `frozenPrefixLen` marks a distinct frozen prefix (the compaction
 * boundary), pin the deep breakpoint there — it is byte-stable across turns,
 * so `[0..frozen]` is re-read cheaply every turn; the tail breakpoint writes
 * only the incremental `[frozen..tail]`. Otherwise fall back to the rolling
 * pair: final message + nearest preceding user turn.
 */
function applyMessageCacheControl(msgs: Array<Record<string, unknown>>, frozenPrefixLen?: number): void {
  if (!msgs.length) return
  const targets = new Set<number>([msgs.length - 1])
  if (typeof frozenPrefixLen === "number" && frozenPrefixLen >= 1 && frozenPrefixLen < msgs.length) {
    targets.add(frozenPrefixLen - 1)
  } else {
    for (let i = msgs.length - 2; i >= 0 && targets.size < MESSAGE_CACHE_BREAKPOINTS; i--) {
      if (msgs[i].role === "user") targets.add(i)
    }
  }
  for (const idx of targets) markLastBlockCacheable(msgs[idx])
}

function markLastBlockCacheable(msg: Record<string, unknown>): void {
  const cache_control = { type: "ephemeral" as const }
  if (typeof msg.content === "string") {
    if (!msg.content) return
    msg.content = [{ type: "text", text: msg.content, cache_control }]
    return
  }
  if (Array.isArray(msg.content) && msg.content.length) {
    const last = msg.content[msg.content.length - 1] as Record<string, unknown>
    last.cache_control = cache_control
  }
}

/** Regression guard: fail loudly before the API would reject the request for
 *  exceeding the cache_control breakpoint limit. */
function assertCacheBudget(system: unknown, toolCount: number): void {
  const systemBreakpoints = Array.isArray(system) ? system.length : 0
  const toolBreakpoints = toolCount > 0 && !Array.isArray(system) ? 1 : 0
  if (systemBreakpoints + toolBreakpoints + MESSAGE_CACHE_BREAKPOINTS > MAX_CACHE_BREAKPOINTS) {
    throw new Error(
      `Anthropic cache_control budget exceeded: ${systemBreakpoints} system + ${toolBreakpoints} tool + ${MESSAGE_CACHE_BREAKPOINTS} message > ${MAX_CACHE_BREAKPOINTS}`,
    )
  }
}

export class AnthropicProvider implements LLMProvider {
  private nativeAssistantBlocks = new Map<string, Array<Record<string, unknown>>>()

  constructor(
    private readonly apiKey: string,
    private readonly model = "claude-sonnet-4-6",
    private readonly maxTokens = 8096,
  ) {}

  descriptor(): ProviderDescriptor {
    return {
      provider: "anthropic",
      protocol: "anthropic-messages",
      model: this.model,
      reasoning: {
        supported: true,
        preserveAcrossToolTurns: true,
        requiresReplayForToolTurns: true,
      },
      toolCalls: {
        supported: true,
        requiresStrictPairing: true,
      },
    }
  }

  peekProviderReplay(message: Pick<Message, "content" | "toolCalls">): ProviderReplay | undefined {
    const blocks = this.nativeAssistantBlocks.get(assistantReplayKey(message))
    return blocks?.length ? { native_blocks: blocks } : undefined
  }

  seedProviderReplay(message: Pick<Message, "content" | "toolCalls">, replay: ProviderReplay): void {
    if (replay.native_blocks?.length) {
      this.nativeAssistantBlocks.set(assistantReplayKey(message), replay.native_blocks)
      return
    }
    // Legacy log without persisted native blocks: reconstruct neutral
    // text + tool_use blocks so a tool-use turn can be replayed.
    const blocks = reconstructAnthropicBlocks(message)
    if (blocks.length) this.nativeAssistantBlocks.set(assistantReplayKey(message), blocks)
  }

  async complete(context: RenderedContext, tools: ToolSchema[], extensions?: Record<string, unknown>): Promise<Message> {
    return collectStreamMessage(this.stream(context, tools, extensions))
  }

  async *stream(context: RenderedContext, tools: ToolSchema[], extensions?: Record<string, unknown>): AsyncIterable<StreamEvent> {
    const systemBlocks: Array<{ type: "text"; text: string; cache_control?: { type: "ephemeral" } }> = []
    if (context.systemStable) {
      systemBlocks.push({ type: "text", text: context.systemStable, cache_control: { type: "ephemeral" } })
    }
    if (context.systemKnowledge) {
      systemBlocks.push({ type: "text", text: context.systemKnowledge, cache_control: { type: "ephemeral" } })
    }
    const system = systemBlocks.length ? systemBlocks : (context.systemText || undefined)
    const msgs = toAnthropicMessages(context, message =>
      this.nativeAssistantBlocks.get(assistantReplayKey(message))
    )
    applyMessageCacheControl(msgs, context.frozenPrefixLen)
    // Append the volatile State turn AFTER the cache breakpoints (uncached tail);
    // absent on un-rebuilt bindings, where the state is already inside `turns`.
    // Render through toAnthropicMessages so assistant tool_use blocks are
    // serialized correctly — raw content would silently drop toolCalls.
    if (context.stateTurn) {
      const stateCtx: RenderedContext = { systemText: "", turns: [context.stateTurn] }
      const stateMsgs = toAnthropicMessages(stateCtx, message =>
        this.nativeAssistantBlocks.get(assistantReplayKey(message))
      )
      msgs.push(...stateMsgs)
    }
    assertCacheBudget(system, tools.length)

    const body: Record<string, unknown> = {
      model: this.model,
      max_tokens: this.maxTokens,
      messages: msgs,
      stream: true,
      ...(system ? { system } : {}),
      ...(tools.length ? { tools: buildAnthropicTools(tools, !Array.isArray(system)) } : {}),
    }
    if (extensions?.enable_thinking) {
      body.thinking = { type: "enabled", budget_tokens: 8000 }
    }

    const resp = await fetch("https://api.anthropic.com/v1/messages", {
      method: "POST",
      headers: {
        "x-api-key": this.apiKey,
        "anthropic-version": "2023-06-01",
        "content-type": "application/json",
        "anthropic-beta": "prompt-caching-2024-07-31",
      },
      body: JSON.stringify(body),
    })
    if (!resp.ok) throw new Error(`Anthropic ${resp.status}: ${await resp.text()}`)

    const toolBlocks: Record<number, { id: string; name: string; argsBuf: string }> = {}
    const nativeBlocks: Record<number, Record<string, unknown>> = {}
    let finalText = ""
    const finalToolCalls: Array<{ id: string; name: string; arguments: string }> = []
    const reader = resp.body!.getReader()
    const decoder = new TextDecoder()
    let buf = ""
    let uncachedInput = 0
    let cacheReadTokens = 0
    let cacheCreationTokens = 0
    let outputTokens = 0

    while (true) {
      const { done, value } = await reader.read()
      if (done) break
      buf += decoder.decode(value, { stream: true })
      const lines = buf.split("\n")
      buf = lines.pop() ?? ""
      for (const line of lines) {
        if (!line.startsWith("data: ")) continue
        const data = line.slice(6).trim()
        if (data === "[DONE]") return
        try {
          const evt = JSON.parse(data) as Record<string, unknown>
          if (evt.type === "message_start" || evt.type === "message_delta") {
            const usage = (evt.usage ?? (evt.message as Record<string, unknown> | undefined)?.usage) as
              | { input_tokens?: number; output_tokens?: number; cache_read_input_tokens?: number; cache_creation_input_tokens?: number }
              | undefined
            if (usage) {
              // input + cache counts are pinned at message_start; a later
              // message_delta may omit them — Math.max prevents zeroing.
              uncachedInput = Math.max(uncachedInput, usage.input_tokens ?? 0)
              cacheReadTokens = Math.max(cacheReadTokens, usage.cache_read_input_tokens ?? 0)
              cacheCreationTokens = Math.max(cacheCreationTokens, usage.cache_creation_input_tokens ?? 0)
              outputTokens = Math.max(outputTokens, usage.output_tokens ?? 0)
              // inputTokens is the FULL prompt (uncached + cache read + write):
              // the kernel reads it as the authoritative context size.
              const inputTokens = uncachedInput + cacheReadTokens + cacheCreationTokens
              if (inputTokens > 0 || outputTokens > 0) {
                yield {
                  type: "usage",
                  totalTokens: inputTokens + outputTokens,
                  inputTokens,
                  outputTokens,
                  cacheReadInputTokens: cacheReadTokens,
                  cacheCreationInputTokens: cacheCreationTokens,
                } as UsageEvent
              }
            }
          } else if (evt.type === "content_block_start") {
            const idx = evt.index as number
            nativeBlocks[idx] = { ...(evt.content_block as Record<string, unknown>) }
            const cb = evt.content_block as Record<string, unknown>
            if (cb.type === "tool_use")
              toolBlocks[idx] = { id: cb.id as string, name: cb.name as string, argsBuf: "" }
          } else if (evt.type === "content_block_delta") {
            const d = evt.delta as Record<string, unknown>
            const idx = evt.index as number
            if (d.type === "text_delta") {
              finalText += String(d.text)
              nativeBlocks[idx] = { ...nativeBlocks[idx], text: String(nativeBlocks[idx]?.text ?? "") + d.text }
              yield { type: "text_delta", delta: d.text } as TextDelta
            } else if (d.type === "thinking_delta") {
              nativeBlocks[idx] = { ...nativeBlocks[idx], thinking: String(nativeBlocks[idx]?.thinking ?? "") + d.thinking }
              yield { type: "thinking_delta", delta: d.thinking } as ThinkingDelta
            } else if (d.type === "signature_delta") {
              nativeBlocks[idx] = { ...nativeBlocks[idx], signature: String(nativeBlocks[idx]?.signature ?? "") + d.signature }
            } else if (d.type === "input_json_delta" && toolBlocks[idx]) {
              toolBlocks[idx].argsBuf += d.partial_json
            }
          } else if (evt.type === "content_block_stop") {
            const idx = evt.index as number
            if (toolBlocks[idx]) {
              const tb = toolBlocks[idx]
              delete toolBlocks[idx]
              let args: Record<string, unknown> = {}
              try { args = JSON.parse(tb.argsBuf || "{}") } catch { args = {} }
              nativeBlocks[idx] = { ...nativeBlocks[idx], input: args }
              finalToolCalls.push({ id: tb.id, name: tb.name, arguments: JSON.stringify(args) })
              yield { type: "tool_call", id: tb.id, name: tb.name, arguments: args } as ToolCallEvent
            }
          }
        } catch { /* skip malformed */ }
      }
    }

    this.rememberNativeBlocks({ content: finalText, toolCalls: finalToolCalls }, Object.keys(nativeBlocks).map(Number).sort((a, b) => a - b).map(index => nativeBlocks[index]))
  }

  private rememberNativeBlocks(
    message: Pick<Message, "content" | "toolCalls">,
    blocks: Array<Record<string, unknown>>,
  ): void {
    if (!blocks.length) return
    if (!message.toolCalls?.length && !blocks.some(b => b.type === "thinking")) return
    this.nativeAssistantBlocks.set(assistantReplayKey(message), blocks)
  }
}

/**
 * Reconstruct Anthropic assistant content blocks from a neutral transcript when
 * no provider replay was persisted. Only meaningful for tool-use turns.
 */
function reconstructAnthropicBlocks(
  message: Pick<Message, "content" | "toolCalls">,
): Array<Record<string, unknown>> {
  const toolCalls = message.toolCalls ?? []
  if (!toolCalls.length) return []
  const blocks: Array<Record<string, unknown>> = []
  if (message.content) blocks.push({ type: "text", text: message.content })
  for (const tc of toolCalls) {
    let input: Record<string, unknown> = {}
    try { input = JSON.parse(tc.arguments || "{}") as Record<string, unknown> } catch { input = {} }
    blocks.push({ type: "tool_use", id: tc.id, name: tc.name, input })
  }
  return blocks
}

File: python/deepstrike/providers/anthropic.py
from __future__ import annotations
import json
import logging
from typing import AsyncIterator
from anthropic import AsyncAnthropic
from deepstrike._kernel import Message, ToolCall, ToolSchema
from .stream import StreamEvent, TextDelta, ThinkingDelta, ToolCallEvent, UsageEvent
from .base import RetryConfig, CircuitBreaker, ProviderDescriptor, RenderedContext, RuntimePolicy, normalize_tool_call, parse_tool_arguments, to_anthropic_content, to_anthropic_messages

logger = logging.getLogger(__name__)

_CLAUDE_POLICIES: dict[str, RuntimePolicy] = {
    "claude-opus-4-1":           RuntimePolicy(max_turns=50),
    "claude-opus-4-7":           RuntimePolicy(max_turns=50),
    "claude-opus-4-6":           RuntimePolicy(max_turns=50),
    "claude-opus-4-0":           RuntimePolicy(max_turns=50),
    "claude-sonnet-4-6":         RuntimePolicy(max_turns=25),
    "claude-sonnet-4-0":         RuntimePolicy(max_turns=25),
    "claude-haiku-4-5":          RuntimePolicy(max_turns=15),
    "claude-3-5-haiku-latest":   RuntimePolicy(max_turns=15),
}


class AnthropicProvider:
    def __init__(
        self,
        api_key: str,
        model: str = "claude-sonnet-4-6",
        retry_config: RetryConfig | None = None,
        base_url: str | None = None,
    ):
        self._model = model
        self._retry = retry_config or RetryConfig()
        self._circuit = CircuitBreaker(self._retry)
        client_kwargs: dict = {"api_key": api_key}
        if base_url:
            client_kwargs["base_url"] = base_url
        self._client = AsyncAnthropic(**client_kwargs)
        self._native_assistant_blocks: dict[str, list[dict]] = {}

    def runtime_policy(self) -> RuntimePolicy:
        return _CLAUDE_POLICIES.get(self._model, RuntimePolicy())

    def _provider_name(self) -> str:
        """Identity advertised in the descriptor; overridden by Anthropic-compatible vendors (e.g. MiniMax)."""
        return "anthropic"

    def descriptor(self) -> ProviderDescriptor:
        return ProviderDescriptor(
            provider=self._provider_name(),
            protocol="anthropic-messages",
            model=self._model,
            reasoning={"supported": True, "preserve_across_tool_turns": True, "requires_replay_for_tool_turns": True},
            tool_calls={"supported": True, "requires_strict_pairing": True},
        )

    def peek_provider_replay(self, content: str, tool_calls: list[ToolCall]) -> dict | None:
        blocks = self._native_assistant_blocks.get(self._assistant_replay_key_parts(content, tool_calls))
        return {"native_blocks": blocks} if blocks else None

    def seed_provider_replay(self, content: str, tool_calls: list[ToolCall], replay: dict) -> None:
        blocks = replay.get("native_blocks")
        if blocks:
            self._native_assistant_blocks[self._assistant_replay_key_parts(content, tool_calls)] = blocks
            return
        # Legacy log without persisted native blocks: reconstruct neutral
        # text + tool_use blocks from the transcript so a tool-use turn can be
        # replayed. Thinking blocks were never persisted and are not recovered.
        reconstructed = _reconstruct_anthropic_blocks(content, tool_calls)
        if reconstructed:
            self._native_assistant_blocks[self._assistant_replay_key_parts(content, tool_calls)] = reconstructed

    def _build_system(self, context: RenderedContext):
        """Structured system blocks with cache_control when the kernel partitioned
        the prompt (system_stable / system_knowledge); else the flat system_text
        string (no cache breakpoint), or None."""
        stable = getattr(context, "system_stable", "") or ""
        knowledge = getattr(context, "system_knowledge", "") or ""
        if not stable and not knowledge:
            return context.system_text or None
        # System cache_control is emitted under "default" and "system-only". Other strategies keep
        # the text-block structure for protocol parity but omit cache_control.
        emit = strategy in ("default", "system-only")
        cc = {"cache_control": {"type": "ephemeral"}} if emit else {}
        blocks: list[dict] = []
        if stable:
            blocks.append({"type": "text", "text": stable, **cc})
        if knowledge:
            blocks.append({"type": "text", "text": knowledge, **cc})
        return blocks or None

    def _build_messages(self, turns: list[Message], state_turn=None, frozen_prefix_len=None, strategy: str = "default") -> list[dict]:
        msgs = to_anthropic_messages(
            turns,
            native_replay=lambda message: self._native_assistant_blocks.get(
                self._assistant_replay_key(message)
            ),
        )
        # Cache breakpoints anchor on the stable history; the volatile State turn
        # is appended AFTER them as the uncached tail. On un-rebuilt bindings
        # state_turn is None and the state is already inside turns (rendered above).
        # frozen_prefix_len (P1-E) pins the deep breakpoint at the compaction
        # boundary; None ⇒ rolling-pair fallback.
        _apply_message_cache_control(msgs, frozen_prefix_len, strategy)
        if state_turn is not None:
            # Render through to_anthropic_messages so assistant tool_use blocks
            # and tool-role tool_result parts are serialized correctly —
            # to_anthropic_content only handles contentParts/content and would
            # silently drop tool_calls.
            state_msgs = to_anthropic_messages(
                [state_turn],
                native_replay=lambda message: self._native_assistant_blocks.get(
                    self._assistant_replay_key(message)
                ),
            )
            msgs.extend(state_msgs)
        return msgs

    def _build_tools(self, tools: list[ToolSchema], anchor_cache: bool, strategy: str = "default") -> list[dict] | None:
        if not tools:
            return None
        defs = [
            {
                "name": t.name,
                "description": t.description,
                "input_schema": json.loads(t.parameters),
            }
            for t in tools
        ]
        # Tool cache_control is emitted under "default" and "tools-only".
        emit_on_last_tool = anchor_cache and strategy in ("default", "tools-only")
        if emit_on_last_tool:
            defs[-1]["cache_control"] = {"type": "ephemeral"}
        return defs

    def _remember_native_blocks(self, content: str, tool_calls: list[ToolCall], blocks: list[dict]) -> None:
        if not blocks:
            return
        if not tool_calls and not any(b.get("type") == "thinking" for b in blocks):
            return
        self._native_assistant_blocks[self._assistant_replay_key_parts(content, tool_calls)] = blocks

    async def complete(self, context: RenderedContext, tools: list[ToolSchema], extensions: dict | None = None) -> Message:
        if self._circuit.is_open():
            raise RuntimeError("Circuit breaker open")

        strategy = _resolve_cache_breakpoint_strategy(extensions)
        msgs = self._build_messages(context.turns, context.state_turn, context.frozen_prefix_len, strategy)
        system = self._build_system(context, strategy)
        tool_defs = self._build_tools(tools, anchor_cache=not isinstance(system, list), strategy=strategy)
        _assert_cache_budget(system, len(tools))

        last_exc = None
        for attempt in range(self._retry.max_retries):
            try:
                request_extensions = {k: v for k, v in (extensions or {}).items() if k not in {"model", "messages", "system", "tools", "stream", "max_tokens"}}
                resp = await self._client.messages.create(
                    **request_extensions,
                    model=self._model,
                    max_tokens=(extensions or {}).get("max_tokens", 8096),
                    system=system,
                    messages=msgs,
                    tools=tool_defs,
                )
                self._circuit.record_success()

                content = ""
                tool_calls: list[ToolCall] = []
                native_blocks: list[dict] = []

                for block in resp.content:
                    if block.type == "text":
                        content += block.text
                        native_blocks.append({"type": "text", "text": block.text})
                    elif block.type == "tool_use":
                        tc = normalize_tool_call(block.id, block.name, block.input)
                        if tc:
                            tool_calls.append(tc)
                        native_blocks.append({
                            "type": "tool_use",
                            "id": block.id,
                            "name": block.name,
                            "input": block.input,
                        })
                    elif block.type == "thinking":
                        native_blocks.append({
                            "type": "thinking",
                            "thinking": block.thinking,
                            "signature": getattr(block, "signature", None),
                        })

                self._remember_native_blocks(content, tool_calls, native_blocks)

                # token_count is the turn total: full prompt (uncached + cache
                # read + cache write) + output, so it stays accurate once caching
                # moves most of the prompt into cache_read.
                usage = resp.usage
                full_input = (
                    (usage.input_tokens or 0)
                    + (getattr(usage, "cache_read_input_tokens", 0) or 0)
                    + (getattr(usage, "cache_creation_input_tokens", 0) or 0)
                )
                return Message(
                    role="assistant",
                    content=content,
                    token_count=full_input + (usage.output_tokens or 0),
                    tool_calls=tool_calls or None,
                )
            except Exception as exc:
                last_exc = exc
                self._circuit.record_failure()
                if attempt < self._retry.max_retries - 1:
                    import asyncio
                    delay = self._retry.base_delay * (2 ** attempt)
                    await asyncio.sleep(delay)

        raise last_exc or RuntimeError("Complete failed")

    async def stream(self, context: RenderedContext, tools: list[ToolSchema], extensions: dict | None = None, state: dict | None = None) -> AsyncIterator[StreamEvent]:
        strategy = _resolve_cache_breakpoint_strategy(extensions)
        msgs = self._build_messages(context.turns, context.state_turn, context.frozen_prefix_len, strategy)
        system = self._build_system(context, strategy)
        tool_defs = self._build_tools(tools, anchor_cache=not isinstance(system, list), strategy=strategy)
        _assert_cache_budget(system, len(tools))

        native_blocks: dict[int, dict] = {}
        tool_blocks: dict[int, dict] = {}
        final_text = ""
        final_tool_calls: list[ToolCall] = []
        uncached_input = 0
        cache_read = 0
        cache_creation = 0
        output_tokens = 0

        request_extensions = {k: v for k, v in (extensions or {}).items() if k not in {"model", "messages", "system", "tools", "stream", "max_tokens"}}
        async with self._client.messages.stream(
            **request_extensions,
            model=self._model,
            max_tokens=(extensions or {}).get("max_tokens", 8096),
            system=system,
            messages=msgs,
            tools=tool_defs,
        ) as stream:
            async for event in stream:
                if event.type in ("message_start", "message_delta"):
                    usage = getattr(event, "usage", None) or getattr(getattr(event, "message", None), "usage", None)
                    if usage is not None:
                        # input + cache counts are pinned at message_start; a later
                        # message_delta may omit them — max() prevents zeroing.
                        uncached_input = max(uncached_input, getattr(usage, "input_tokens", 0) or 0)
                        cache_read = max(cache_read, getattr(usage, "cache_read_input_tokens", 0) or 0)
                        cache_creation = max(cache_creation, getattr(usage, "cache_creation_input_tokens", 0) or 0)
                        output_tokens = max(output_tokens, getattr(usage, "output_tokens", 0) or 0)
                        # input_tokens is the FULL prompt size (uncached + cache
                        # read + cache write) for accurate context accounting.
                        full_input = uncached_input + cache_read + cache_creation
                        yield UsageEvent(
                            total_tokens=full_input + output_tokens,
                            input_tokens=full_input,
                            output_tokens=output_tokens,
                            cache_read_input_tokens=cache_read,
                            cache_creation_input_tokens=cache_creation,
                        )
                elif event.type == "content_block_start":
                    idx = event.index
                    block = event.content_block
                    native_blocks[idx] = {"type": block.type}
                    if block.type == "thinking":
                        native_blocks[idx]["thinking"] = getattr(block, "thinking", "") or ""
                        native_blocks[idx]["signature"] = getattr(block, "signature", "") or ""
                    elif block.type == "text":
                        native_blocks[idx]["text"] = getattr(block, "text", "") or ""
                    elif block.type == "tool_use":
                        native_blocks[idx].update({
                            "id": block.id,
                            "name": block.name,
                            "input": getattr(block, "input", {}) or {},
                        })
                        tool_blocks[idx] = {"id": block.id, "name": block.name, "args_buf": ""}
                elif event.type == "content_block_delta":
                    delta = event.delta
                    idx = event.index
                    if delta.type == "text_delta":
                        final_text += delta.text
                        native_blocks[idx]["text"] = native_blocks[idx].get("text", "") + delta.text
                        yield TextDelta(delta=delta.text)
                    elif delta.type == "thinking_delta":
                        native_blocks[idx]["thinking"] = native_blocks[idx].get("thinking", "") + delta.thinking
                        yield ThinkingDelta(delta=delta.thinking)
                    elif delta.type == "signature_delta":
                        native_blocks[idx]["signature"] = native_blocks[idx].get("signature", "") + delta.signature
                    elif delta.type == "input_json_delta" and idx in tool_blocks:
                        tool_blocks[idx]["args_buf"] += delta.partial_json
                elif event.type == "content_block_stop":
                    idx = event.index
                    if idx in tool_blocks:
                        tb = tool_blocks.pop(idx)
                        try:
                            args = json.loads(tb["args_buf"] or "{}")
                        except json.JSONDecodeError:
                            args = {}
                        native_blocks[idx]["input"] = args
                        tc = normalize_tool_call(tb["id"], tb["name"], args)
                        if tc:
                            final_tool_calls.append(tc)
                            yield ToolCallEvent(id=tc.id, name=tc.name, arguments=args)

        ordered_blocks = [native_blocks[i] for i in sorted(native_blocks)]
        self._remember_native_blocks(final_text, final_tool_calls, ordered_blocks)

    def _assistant_replay_key(self, message: Message) -> str:
        return self._assistant_replay_key_parts(message.content, message.tool_calls or [])

    def _assistant_replay_key_parts(self, content: str, tool_calls: list[ToolCall]) -> str:
        return json.dumps({
            "content": content,
            "tool_calls": [
                {"id": _tc_field(tc, "id"), "name": _tc_field(tc, "name"), "arguments": _tc_field(tc, "arguments")}
                for tc in tool_calls
            ],
        }, sort_keys=True)


# Anthropic accepts at most 4 cache_control breakpoints per request; the static
# system/tools prefix uses up to 2, leaving 2 for the rolling message history.
_MAX_CACHE_BREAKPOINTS = 4
_MESSAGE_CACHE_BREAKPOINTS = 2

# Recognised values for the `cacheBreakpointStrategy` extension. See node/src/types.ts
# `CacheBreakpointStrategy` for the canonical documentation; this mirrors the same surface.
_CACHE_BREAKPOINT_STRATEGIES = {"default", "tools-only", "system-only", "frozen-prefix", "none"}


def _resolve_cache_breakpoint_strategy(extensions: "dict | None") -> str:
    """Pull `cacheBreakpointStrategy` from per-call extensions; unrecognised values → 'default'."""
    if not extensions:
        return "default"
    raw = extensions.get("cacheBreakpointStrategy")
    if isinstance(raw, str) and raw in _CACHE_BREAKPOINT_STRATEGIES:
        return raw
    return "default"


def _apply_message_cache_control(msgs: list[dict], frozen_prefix_len: "int | None" = None, strategy: str = "default") -> None:
    """Place the (<=2) message-history cache breakpoints. The final message always
    gets one (writes the current full prefix). The second is placed by one of two
    strategies:

      * Deep anchor (P1-E): when ``frozen_prefix_len`` marks a distinct frozen prefix
        (the compaction boundary), pin the second breakpoint there. It is byte-stable
        across turns, so ``[0..frozen]`` is re-read cheaply every turn and is immune to
        the 20-block lookback miss on heavy tool turns; the tail then writes only the
        incremental ``[frozen..tail]``.
      * Rolling fallback: otherwise mark the nearest preceding user turn (the previous
        turn's read anchor); Anthropic's 20-block lookback bridges light turns.

    Without this the cached prefix stops at ``system`` and the whole tool-result history
    is re-billed at full price every turn (~quadratic cumulative cost). A bare string
    body is promoted to a cache-bearing text block."""
    if not msgs:
        return
    # Message-level cache_control is emitted under "default" and "frozen-prefix" only.
    if strategy in ("tools-only", "system-only", "none"):
        return
    targets = {len(msgs) - 1}
    if isinstance(frozen_prefix_len, int) and 1 <= frozen_prefix_len < len(msgs):
        # Deep anchor at the frozen-prefix boundary (last frozen turn). Fixed between compactions.
        targets.add(frozen_prefix_len - 1)
    elif strategy == "default":
        i = len(msgs) - 2
        while i >= 0 and len(targets) < _MESSAGE_CACHE_BREAKPOINTS:
            if msgs[i].get("role") == "user":
                targets.add(i)
            i -= 1
    for idx in targets:
        _mark_last_block_cacheable(msgs[idx])


def _mark_last_block_cacheable(msg: dict) -> None:
    cache_control = {"type": "ephemeral"}
    content = msg.get("content")
    if isinstance(content, str):
        if not content:
            return  # don't synthesize an empty (API-rejected) text block
        msg["content"] = [{"type": "text", "text": content, "cache_control": cache_control}]
        return
    if isinstance(content, list) and content:
        content[-1]["cache_control"] = cache_control


def _assert_cache_budget(system: object, tool_count: int) -> None:
    """Regression guard: fail loudly before the API would reject the request for
    exceeding the cache_control breakpoint limit."""
    system_breakpoints = len(system) if isinstance(system, list) else 0
    tool_breakpoints = 1 if tool_count > 0 and not isinstance(system, list) else 0
    worst_case = system_breakpoints + tool_breakpoints + _MESSAGE_CACHE_BREAKPOINTS
    if worst_case > _MAX_CACHE_BREAKPOINTS:
        raise ValueError(
            f"Anthropic cache_control budget exceeded: {system_breakpoints} system + "
            f"{tool_breakpoints} tool + {_MESSAGE_CACHE_BREAKPOINTS} message > {_MAX_CACHE_BREAKPOINTS}"
        )


def _tc_field(tc: object, field_name: str) -> object:
    if isinstance(tc, dict):
        return tc.get(field_name)
    return getattr(tc, field_name, None)


def _reconstruct_anthropic_blocks(content: str, tool_calls: list) -> list[dict]:
    """Reconstruct Anthropic assistant content blocks from a neutral transcript
    when no provider replay was persisted. Only meaningful for tool-use turns."""
    if not tool_calls:
        return []
    blocks: list[dict] = []
    if content:
        blocks.append({"type": "text", "text": content})
    for tc in tool_calls:
        blocks.append({
            "type": "tool_use",
            "id": _tc_field(tc, "id"),
            "name": _tc_field(tc, "name"),
            "input": parse_tool_arguments(_tc_field(tc, "arguments")),
        })
    return blocks
